src/pages/workspaces/workspace/RequestAccessModal.js

```diff
@@ -0,0 +1,103 @@
+
+
+export const RequestAccessModal = ({ onDismiss, workspace }) => {
+  const [groups, setGroups] = useState([])
+  const [accessInstructions, setAccessInstructions] = useState([])
+  const [loadingGroups, setLoadingGroups] = useState(false)
+  const [loadingAccessInstructions, setLoadingAccessInstructions] = useState(false)
+  const signal = useCancellation()
+
+  const { Groups, Workspaces } = Ajax(signal)
+
+  const fetchGroups = _.flow(
+    withBusyState(setLoadingGroups),
+    withErrorReporting('Error loading groups')
+  )(async () => {
+    setGroups(await Groups.list())
+  })
+
+  const fetchAccessInstructions = _.flow(
+    withBusyState(setLoadingAccessInstructions),
+    withErrorReporting('Error loading groups')
+  )(async () => {
+    const i = await Workspaces.workspace(workspace.workspace.namespace, workspace.workspace.name).accessInstructions()
+    console.log(i)
+    setAccessInstructions(i)
+  })
+
+  useOnMount(() => {
+    fetchGroups()
+    fetchAccessInstructions()
+  })
+
+  const loading = loadingGroups || loadingAccessInstructions
+  const groupNames = _.map('groupName', groups)
+  const authDomain = workspace.workspace.authorizationDomain
+  return h(Modal, {
+    title: 'Request Access',
+    width: '40rem',
+    showCancel: false,
+    onDismiss
+  }, [
+    div(['You cannot access this Workspace because it contains restricted data. You need permission from the admin(s) of all of the Groups in the Authorization Domain protecting the workspace.']),
+    loading ?
+      centeredSpinner({ size: 32 }) :
+      table({ style: { margin: '1rem', width: '100%' } }, [
+        thead([
+          tr({ style: { height: '2rem' } }, [th({ style: { textAlign: 'left' } }, ['Group Name']), th({ style: { textAlign: 'left', width: '15rem' } }, ['Access'])])
+        ]),
+        tbody(
+          _.map(({ membersGroupName: groupName }) => tr({ style: { height: '2rem' } }, [
+            td([groupName]),
+            td([
+              _.includes(groupName, groupNames) ?
+                'Yes' :
+                h(RequestAccessButton, {
+                  groupName,
+                  instructions: accessInstructions[groupName]
+                })
+            ])
+          ]), authDomain)
+        )
+      ])
+  ])
+}
+
+const RequestAccessButton = ({ groupName, instructions }) => {
+  const [requesting, setRequesting] = useState(false)
+  const [requested, setRequested] = useState(false)
+  const signal = useCancellation()
+
+  const { Groups } = Ajax(signal)
+
+  const requestAccess = _.flow(
+    withBusyState(setRequesting),
+    withErrorReporting('Error requesting group access')
+  )(async () => {
+    await Groups.group(groupName).requestAccess()
+    setRequested(true)
+  })
+
+  return buttonPrimary({
+    disabled: requesting || requested,
+    onClick: async () => {
+      await requestAccess()
+      setRequested(true)
+    }
+  }, [
+    cond(
+      [requested, 'Request Sent'],
```

Comment:
I like this feedback! And that it's disabled.

Reply:
It's worth noting that we don't persist the fact that someone requested access. Therefore, if you request access and then close and reopen the modal, the button will be enabled again. There's nothing we can do about this without making some changes to some services (probably Sam).

Reply:
That is good to know. Can a user requests access to a group multiple times? Or does Sam prevent multiple requests to/from the same user(s)? To make sure admins aren't spammed with requests by impatient users...

Reply:
Nothing preventing people from spamming admins, AFAIK.

Reply:
Okay, that might be something we have to prevent against in the future, if it becomes a problem.